sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb

```diff
@@ -146,7 +146,7 @@ def shutdown(timeout: nil)
 
             thread&.join(timeout)
             force_flush(timeout: OpenTelemetry::Common::Utilities.maybe_timeout(timeout, start_time))
-            dropped_spans = lock { spans.size }
+            dropped_spans = lock { spans}
```

Comment:
This is problematic if multiple threads are running. We know the exporter thread has finished, and we _should_ be the only thread executing `shutdown`, but there is some possibility that other application threads are running and enqueuing finished spans.

Technically, we haven't "dropped spans" because we're just assuming that the process is shutting down and the BSP will go away anyway, but I think we should not make that assumption. Instead, we should _actually_ drop the spans by shifting them out of the `spans` array. That will also satisfy mutual exclusion concerns.

So, something like:
```suggestion
            dropped_spans = lock { spans.shift(spans.length) }
```

Reply:
> Technically, we haven't "dropped spans" because we're just assuming that the process is shutting down and the BSP will go away anyway, but I think we should not make that assumption. Instead, we should _actually_ drop the spans by shifting them out of the `spans` array. That will also satisfy mutual exclusion concerns.

I think this is a good idea, but also, I'm not clear on why we shouldn't assume the BSP will go away anyway. I was under the impression that there was one BSP worker thread per forked process 🤔

Reply:
closed in favor of https://github.com/open-telemetry/opentelemetry-ruby/pull/1566; we can continue this discussion here or there :)